src/adapter/src/explain/insights.rs

```diff
@@ -28,6 +56,45 @@ pub struct PlanInsights {
     /// Each key is the ID of an imported collection, and each value contains
     /// further insights about each collection and how it is used by the plan.
     pub imports: BTreeMap<String, ImportInsights>,
+    /// If this plan is not fast path, this is the set of cluster names that would render this as fast path.
+    pub fast_path_clusters: BTreeSet<String>,
+}
+
+impl PlanInsights {
+    pub fn compute_fast_path_clusters(&mut self, ctx: PlanInsightsContext) {
+        for (name, compute_instance) in ctx.compute_instances {
+            let mut optimizer = optimize::peek::Optimizer::new(
+                Arc::clone(&ctx.catalog),
+                compute_instance,
+                ctx.finishing.clone(),
+                ctx.view_id,
+                ctx.index_id,
+                ctx.optimizer_config.clone(),
+                ctx.metrics.clone(),
+            );
+
+            let res: Result<_, OptimizerError> = (|| {
+                // HIR ⇒ MIR lowering and MIR optimization (local)
+                let local_mir_plan = optimizer.catch_unwind_optimize(ctx.raw_expr.clone())?;
+                // Attach resolved context required to continue the pipeline.
+                let local_mir_plan = local_mir_plan.resolve(
+                    ctx.timestamp_context.clone(),
+                    &ctx.session,
+                    Box::new(EmptyStatisticsOracle {}),
+                );
+                // MIR optimization (global), MIR ⇒ LIR lowering, and LIR optimization (global)
+                let global_lir_plan = optimizer.catch_unwind_optimize(local_mir_plan)?;
+                Ok(global_lir_plan)
+            })();
+            let Ok(plan) = res else {
+                continue;
+            };
+            let (plan, _, _) = plan.unapply();
+            if matches!(plan, PeekPlan::FastPath(_)) {
+                self.fast_path_clusters.insert(name);
+            }
+        }
```

Comment:
It might be worth optimizing in parallel since sometimes optimization can take awhile.

Reply:
Done. This made a bunch of stuff async but that seems fine.